src/approaches/grammar_search_invention_approach.py

```diff
@@ -802,22 +810,38 @@ class _HeuristicCountBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint
     the transition is optimal and the sequence is not a demo, that's
     assumed to be bad; if the transition is not optimal and the sequence
     is a demo, that's also assumed to be bad.
+
+    Also: for each successor that is one step off the atoms_sequence, if the
+    state is optimal, then check if the state is "suspicious", meaning that it
+    does not "match" any state in the demo data. The definition of match is
+    currently based on utils.unify(). It may be that this definition is too
+    strong, so we could consider others. The idea is to try to distinguish
+    states that are actually impossible (suspicious) from ones that are simply
+    alternative steps toward optimally achieving the goal.
     """
 
-    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
-                                  heuristic_fn: Callable[[Set[GroundAtom]],
-                                                         float],
-                                  ground_ops: Set[_GroundSTRIPSOperator],
-                                  is_demo: bool) -> float:
+    def _evaluate_atom_trajectory(
+        self,
+        atoms_sequence: List[Set[GroundAtom]],
+        heuristic_fn: Callable[[Set[GroundAtom]], float],
+        ground_ops: Set[_GroundSTRIPSOperator],
+        demo_atom_sets: Set[FrozenSet[GroundAtom]],
+        is_demo: bool,
+    ) -> float:
         score = 0.0
         for i in range(len(atoms_sequence) - 1):
             atoms, next_atoms = atoms_sequence[i], atoms_sequence[i + 1]
             best_h = float("inf")
             on_sequence_h = float("inf")
+            optimal_successors = set()
             for predicted_next_atoms in utils.get_successors_from_ground_ops(
                     atoms, ground_ops, unique=False):
                 # Compute the heuristic for the successor atoms.
                 h = heuristic_fn(predicted_next_atoms)
+                if h < best_h:
```

Comment:
I'm confused how the logic is working here. Won't best_h only be guaranteed to be set correctly once the `get_successors_from_ground_ops` loop has gone all the way through? Why is it okay to compare to a best_h that you're still working on figuring out the value of?